voter/src/test/java/org/terracotta/voter/ActiveVoterTest.java

```diff
@@ -0,0 +1,167 @@
+package org.terracotta.voter;
+
+import org.junit.Test;
+
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Function;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.terracotta.voter.ActiveVoter.TOPOLOGY_FETCH_TIME_PROPERTY;
+
+public class ActiveVoterTest {
+  private static final long TOPOLOGY_FETCH_INTERVAL = 11000L;
+
+  static {
+    System.setProperty(TOPOLOGY_FETCH_TIME_PROPERTY, "10000");
+  }
+
+  @Test
+  public void testTopologyUpdate() throws TimeoutException, InterruptedException {
+
+    Set<String> expectedTopology = new HashSet<>();
+    expectedTopology.add("localhost:1234");
+    expectedTopology.add("localhost:1235");
+
+    ClientVoterManager firstClientVoterManager = mock(ClientVoterManager.class);
+    ClientVoterManager otherClientVoterManager = mock(ClientVoterManager.class);
+    when(firstClientVoterManager.isConnected()).thenReturn(true);
+    when(otherClientVoterManager.isConnected()).thenReturn(true);
+    when(firstClientVoterManager.getServerState()).thenReturn("ACTIVE-COORDINATOR");
+    when(otherClientVoterManager.getServerState()).thenReturn("PASSIVE-STANDBY");
+    when(firstClientVoterManager.registerVoter("mvoter")).thenReturn(0L);
+    when(otherClientVoterManager.registerVoter("mvoter")).thenReturn(0L);
+    when(firstClientVoterManager.getTopology()).thenReturn(expectedTopology);
+    when(firstClientVoterManager.heartbeat("mvoter")).thenReturn(0L);
+    when(otherClientVoterManager.heartbeat("mvoter")).thenReturn(0L);
+    when(firstClientVoterManager.getTargetHostPort()).thenReturn("localhost:1234");
+    when(otherClientVoterManager.getTargetHostPort()).thenReturn("localhost:1235");
+
+    Function<String, ClientVoterManager> factory = hostPort -> {
+      if (hostPort.equals("localhost:1234")) {
+        return firstClientVoterManager;
+      } else if (hostPort.equals("localhost:1235")) {
+        return otherClientVoterManager;
+      } else {
+        return new ClientVoterManagerImpl(hostPort);
+      }
+    };
+    ActiveVoter activeVoter = new ActiveVoter("mvoter",
+        new CompletableFuture<VoterStatus>(), Optional.empty(), factory, "localhost:1234", "localhost:1235");
+    activeVoter.start();
+
+    Thread.sleep(TOPOLOGY_FETCH_INTERVAL);
+    System.out.println(activeVoter.getHeartbeatThreads().size() + "->" + activeVoter.getExistingTopology());
+    assertThat(activeVoter.getExistingTopology(), is(expectedTopology));
+    assertThat(activeVoter.getHeartbeatThreads().size(), is(3));
```

Comment:
should be 2, right?

Reply:
3 two for client voters and one for polling

Reply:
we planned to remove the polling thread, right?